rust/opendp/src/trans/variance.rs

```diff
@@ -0,0 +1,175 @@
+    where MI: DatasetMetric<Distance=u32>,
+          MO: SensitivityMetric<Distance=T>,
+          T: 'static + Clone + PartialOrd + Sub<Output=T> + Mul<Output=T> + Div<Output=T> + DistanceCast + Float + Sum<T> + for<'a> Sum<&'a T>,
+          for<'a> &'a T: Sub<Output=T>,
+          Self: BoundedVarianceConstant<MI, MO> {
+    fn make4(lower: T, upper: T, length: usize, ddof: usize) -> Fallible<Transformation<SizedDomain<VectorDomain<IntervalDomain<T>>>, AllDomain<T>, MI, MO>> {
+        if lower > upper { return fallible!(MakeTransformation, "lower bound may not be greater than upper bound"); }
+        let _length = T::from(length).ok_or_else(|| err!(FailedCast))?;
+        let _ddof = T::from(ddof).ok_or_else(|| err!(FailedCast))?;
+
+        Ok(Transformation::new(
+            SizedDomain::new(VectorDomain::new(
+                IntervalDomain::new(Bound::Included(lower.clone()), Bound::Included(upper.clone()))), length),
+            AllDomain::new(),
+            Function::new(move |arg: &Vec<T>| {
+                let mean = arg.iter().sum::<T>() / _length;
+                arg.iter().map(|v| (v - &mean).powi(2)).sum::<T>() / (_length - _ddof)
+            }),
+            MI::new(),
+            MO::new(),
+            StabilityRelation::new_from_constant(Self::get_stability(lower, upper, length, ddof)?)))
+    }
+}
+
+
+pub struct BoundedCovariance<MI, MO> {
+    input_metric: PhantomData<MI>,
+    output_metric: PhantomData<MO>,
+}
+
+
+pub trait BoundedCovarianceConstant<MI: Metric, MO: Metric> {
+    fn get_stability_constant(lower: (MO::Distance, MO::Distance), upper: (MO::Distance, MO::Distance), length: usize, ddof: usize) -> Fallible<MO::Distance>;
+}
+
+impl<MO: Metric<Distance=T>, T> BoundedCovarianceConstant<HammingDistance, MO> for BoundedCovariance<HammingDistance, MO>
+    where T: Clone + Sub<Output=T> + Div<Output=T> + NumCast + One {
+    fn get_stability_constant(lower: (T, T), upper: (T, T), length: usize, ddof: usize) -> Fallible<T> {
+        let _length = T::from(length).ok_or_else(|| err!(FailedCast))?;
+        let _1 = T::one();
+        let _ddof = T::from(ddof).ok_or_else(|| err!(FailedCast))?;
+        Ok((upper.0 - lower.0) * (upper.1 - lower.1) * (_length.clone() - _1) / _length.clone() / (_length - _ddof))
+    }
+}
+
+impl<MO: Metric<Distance=T>, T> BoundedCovarianceConstant<SymmetricDistance, MO> for BoundedCovariance<SymmetricDistance, MO>
+    where T: Clone + Sub<Output=T> + Div<Output=T> + Add<Output=T> + NumCast + One {
+    fn get_stability_constant(lower: (T, T), upper: (T, T), length: usize, ddof: usize) -> Fallible<T> {
+        let _length = T::from(length).ok_or_else(|| err!(FailedCast))?;
+        let _1 = T::one();
+        let _ddof = T::from(ddof).ok_or_else(|| err!(FailedCast))?;
+        Ok((upper.0 - lower.0) * (upper.1 - lower.1) * _length.clone() / (_length.clone() + _1) / (_length - _ddof))
+    }
+}
+
+type CovarianceDomain<T> = SizedDomain<VectorDomain<IntervalDomain<(T, T)>>>;
```

Comment:
It's a little sad that we don't have a way to generate data in this shape from Python, but I think this is fine until we settle on a general representation for dataframe-like things (Arrow or whatever).